pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,289 @@
+
+	if time.Since(getBackupCompletedTime(newestBackup)) > backupIntervalHours {
+		newBackup, err := c.createNewBackup(cluster)
+		if err != nil {
+			return fmt.Errorf("[etcd-backup] Backup create failed:%v", err)
+		}
+		logrus.Infof("[etcd-backup] New backup created: %s", newBackup.Name)
+	}
+
+	// rotate old backups
+	retention := cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.Retention
+
+	expiredBackups := getExpiredBackups(retention, internvalHours, clusterBackups)
+	for _, backup := range expiredBackups {
+		if err := c.backupClient.DeleteNamespaced(backup.Namespace, backup.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) createNewBackup(cluster *v3.Cluster) (*v3.EtcdBackup, error) {
+	newBackup := NewBackupObject(cluster)
+	v3.BackupConditionCreated.CreateUnknownIfNotExists(newBackup)
+	return c.backupClient.Create(newBackup)
+
+}
+
+func NewBackupObject(cluster *v3.Cluster) *v3.EtcdBackup {
+	controller := true
+	return &v3.EtcdBackup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    cluster.Name,
+			GenerateName: fmt.Sprintf("%s-", cluster.Name),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+					APIVersion: cluster.APIVersion,
+					Kind:       cluster.Kind,
+					Controller: &controller,
+				},
+			},
+		},
+		ClusterID: cluster.Name,
+	}
+}
+
+func getBackupFilename(snapshotName string, cluster *v3.Cluster) string {
+	if cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig == nil ||
+		cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig == nil {
+		return ""
+	}
+	target := cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s/%s", target.Region, target.BucketName, snapshotName)
+}
+
+func (c *Controller) deleteS3Snapshot(b *v3.EtcdBackup) error {
+	if b.BackupConfig.S3BackupConfig == nil {
+		return fmt.Errorf("can't find S3 backup target configuration")
+	}
+	var err error
+	s3Client := &minio.Client{}
+	endpoint := b.BackupConfig.S3BackupConfig.Endpoint
+	bucket := b.BackupConfig.S3BackupConfig.BucketName
+	// no access credentials, we assume IAM roles
+	if b.BackupConfig.S3BackupConfig.AccessKey == "" ||
+		b.BackupConfig.S3BackupConfig.SecretKey == "" {
+		creds := credentials.NewIAM("")
+		s3Client, err = minio.NewWithCredentials(endpoint, creds, true, "")
+	} else {
+		accessKey := b.BackupConfig.S3BackupConfig.AccessKey
+		secretKey := b.BackupConfig.S3BackupConfig.SecretKey
+
+		s3Client, err = minio.New(endpoint, accessKey, secretKey, true)
+	}
+	if err != nil {
+		return err
+	}
+
+	exists, err := s3Client.BucketExists(bucket)
+	if err != nil {
+		return fmt.Errorf("can't access bucket: %v", err)
+	}
+	if !exists {
+		return fmt.Errorf("bucket %s doesn't exist", bucket)
```

Comment:
do we have to return an error if doesn't exist? our intent is to remove it, so we should just log it in info, and return